File: src/bearings/web/models/vault.py
# mypy: disable-error-code=explicit-any
"""Pydantic wire shapes for ``routes/vault.py``.

Per ``docs/architecture-v1.md`` §1.1.5 the wire DTOs live alongside
the route module. The shapes mirror :class:`bearings.db.vault.VaultEntry`
+ :class:`bearings.agent.vault.SearchHit` /
:class:`bearings.agent.vault.SearchResult` /
:class:`bearings.agent.vault.Redaction` row dataclasses; the route
constructs an ``Out`` model from a row dataclass at the wire boundary.

The ``mypy: disable-error-code=explicit-any`` pragma matches the same
narrow carve-out :mod:`bearings.web.models.tags` makes for Pydantic's
metaclass-exposed ``Any`` surface.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class VaultEntryOut(BaseModel):
    """Response shape for a single vault row.

    Per ``docs/behavior/vault.md`` §"When the user opens the vault"
    each row carries an absolute path, slug, optional title, kind,
    mtime, and size; per §"Paste-into-message behavior" the
    ``markdown_link`` field is server-computed so the client doesn't
    have to assemble it (matches the ``group`` server-computed field
    pattern on :class:`bearings.web.models.tags.TagOut`).
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    path: str
    slug: str
    title: str | None
    kind: str
    mtime: int
    size: int
    last_indexed_at: int
    markdown_link: str


class VaultListOut(BaseModel):
    """Response shape for ``GET /api/vault`` — bucketed list.

    Per vault.md §"When the user opens the vault" the user sees a
    Plans section and a Todos section; the wire surfaces both
    buckets as separate arrays so the client doesn't have to filter
    on the wire-side ``kind`` discriminator. ``plan_roots`` /
    ``todo_globs`` mirror the configured paths so the empty-state
    UI ("No plans found under …") can name them per vault.md
    §"When the user opens the vault" → empty state.
    """

    model_config = ConfigDict(extra="forbid")

    plans: list[VaultEntryOut]
    todos: list[VaultEntryOut]
    plan_roots: list[str]
    todo_globs: list[str]


class RedactionOut(BaseModel):
    """One redaction range. Mirrors :class:`bearings.agent.vault.Redaction`."""

    model_config = ConfigDict(extra="forbid")

    offset: int
    length: int
    pattern: str


class VaultDocOut(BaseModel):
    """Response shape for ``GET /api/vault/{id}`` — entry + body + redactions.

    Per vault.md §"Redaction rendering" the server returns the raw
    body plus a list of redaction ranges; the client masks visually
    and toggles per-range. ``persists no toggle state`` — the server
    re-detects on every fetch (so a file edited to remove a secret
    surfaces unredacted on the next read).
    """

    model_config = ConfigDict(extra="forbid")

    entry: VaultEntryOut
    body: str
    redactions: list[RedactionOut]
    truncated: bool


class SearchHitOut(BaseModel):
    """One search hit row. Mirrors :class:`bearings.agent.vault.SearchHit`."""

    model_config = ConfigDict(extra="forbid")

    vault_id: int
    path: str
    title: str | None
    kind: str
    line_number: int
    snippet: str


class SearchResultOut(BaseModel):
    """Response shape for ``GET /api/vault/search``.

    ``capped`` reflects the "showing first N — narrow your query for
    more" indicator from vault.md §"Search semantics".
    """

    model_config = ConfigDict(extra="forbid")

    hits: list[SearchHitOut]
    capped: bool


__all__ = [
    "RedactionOut",
    "SearchHitOut",
    "SearchResultOut",
    "VaultDocOut",
    "VaultEntryOut",
    "VaultListOut",
]

File: frontend/src/lib/api/vault.ts
/**
 * Typed client for the read-only vault surface (item 1.5;
 * ``src/bearings/web/routes/vault.py``).
 *
 * Per ``docs/behavior/vault.md`` the vault is a read-only filesystem
 * index over ``~/.claude/plans/*.md`` + project ``TODO.md`` files —
 * the user can list, open, search, and copy out, but cannot create /
 * edit / rename / delete. This module surfaces the four observation
 * endpoints and intentionally exposes no write helpers so a
 * coding-standards review can grep this file and verify the read-only
 * semantic at the wire boundary.
 *
 * Wire shapes mirror the Pydantic models in
 * :mod:`bearings.web.models.vault`. A field rename on the backend MUST
 * be reflected here in the same commit; the route handlers use
 * ``extra="forbid"`` so a stray TS field also surfaces at the wire
 * boundary as a 422.
 */
import {
  API_VAULT_ENDPOINT,
  API_VAULT_BY_PATH_ENDPOINT,
  API_VAULT_SEARCH_ENDPOINT,
  vaultDocEndpoint,
  type VaultKind,
} from "../config";
import { getJson, type RequestOptions } from "./client";

/**
 * Wire shape for one vault row — one-to-one with
 * :class:`bearings.web.models.vault.VaultEntryOut`.
 *
 * ``markdown_link`` is server-computed (assembled by
 * :func:`bearings.agent.vault.build_markdown_link`) so the client
 * doesn't have to re-derive ``[Title](file:///abs/path)`` and risk
 * drifting from the linkifier's escape rules — see vault.md
 * §"Paste-into-message behavior".
 */
export interface VaultEntryOut {
  id: number;
  path: string;
  slug: string;
  title: string | null;
  /** Mirrors :const:`VAULT_KIND_PLAN` / :const:`VAULT_KIND_TODO`. */
  kind: VaultKind;
  mtime: number;
  size: number;
  last_indexed_at: number;
  markdown_link: string;
}

/**
 * Bucketed list shape — one-to-one with
 * :class:`bearings.web.models.vault.VaultListOut`. ``plan_roots`` /
 * ``todo_globs`` round-trip the configured paths so the empty-state
 * UI can name them per vault.md §"empty state" — "No plans found
 * under <configured roots>".
 */
export interface VaultListOut {
  plans: VaultEntryOut[];
  todos: VaultEntryOut[];
  plan_roots: string[];
  todo_globs: string[];
}

/**
 * One redaction range — mirrors :class:`bearings.web.models.vault.
 * RedactionOut`. The reading panel masks the slice ``body[offset,
 * offset + length)`` until the user toggles Show.
 */
export interface RedactionOut {
  offset: number;
  length: number;
  /** Detection pattern label (e.g. ``"key=value"``); useful for tooltips. */
  pattern: string;
}

/**
 * Single-doc shape — mirrors :class:`bearings.web.models.vault.
 * VaultDocOut`. ``body`` is the raw markdown; ``redactions`` is the
 * list of byte ranges the renderer should mask. ``truncated`` is
 * ``true`` when the on-disk size exceeded the server-side cap and the
 * tail was elided (vault.md §"Failure modes" — "Read error on a
 * single doc").
 */
export interface VaultDocOut {
  entry: VaultEntryOut;
  body: string;
  redactions: RedactionOut[];
  truncated: boolean;
}

/**
 * One search hit — mirrors :class:`bearings.web.models.vault.
 * SearchHitOut`. ``vault_id`` lets the UI cross-reference back into
 * the bucketed list when the user clicks a hit.
 */
export interface SearchHitOut {
  vault_id: number;
  path: string;
  title: string | null;
  kind: VaultKind;
  line_number: number;
  snippet: string;
}

/**
 * Search-result envelope. ``capped`` is the truthy flag the UI uses
 * to surface the "showing first N — narrow your query" indicator
 * per vault.md §"Search semantics".
 */
export interface SearchResultOut {
  hits: SearchHitOut[];
  capped: boolean;
}

/**
 * ``GET /api/vault`` — re-scans the filesystem and returns the
 * bucketed list. Per vault.md §"Failure modes" the index re-scans on
 * every list request, so callers don't need a separate "refresh"
 * action — just call :func:`listVault` again.
 */
export async function listVault(options: RequestOptions = {}): Promise<VaultListOut> {
  return await getJson<VaultListOut>(API_VAULT_ENDPOINT, options);
}

/**
 * ``GET /api/vault/search?q=...`` — case-insensitive substring
 * search. Whitespace-only queries return an empty result on the
 * server; the client passes them through verbatim and the empty-
 * state copy renders.
 */
export async function searchVault(
  query: string,
  options: RequestOptions = {},
): Promise<SearchResultOut> {
  const merged: RequestOptions = { ...options, query: [["q", query]] };
  return await getJson<SearchResultOut>(API_VAULT_SEARCH_ENDPOINT, merged);
}

/**
 * ``GET /api/vault/{id}`` — open by cache id. Read errors return
 * an empty body + empty redactions per vault.md §"Failure modes" —
 * "Read error on a single doc"; the entry metadata still surfaces.
 */
export async function getVaultDoc(
  vaultId: number,
  options: RequestOptions = {},
): Promise<VaultDocOut> {
  return await getJson<VaultDocOut>(vaultDocEndpoint(vaultId), options);
}

/**
 * ``GET /api/vault/by-path?path=...`` — open by absolute path.
 * The server resolves symlinks before the allowlist check (vault.md
 * §"Failure modes" — "Path outside the vault"); a path that resolves
 * outside the configured roots returns 404.
 */
export async function getVaultDocByPath(
  path: string,
  options: RequestOptions = {},
): Promise<VaultDocOut> {
  const merged: RequestOptions = { ...options, query: [["path", path]] };
  return await getJson<VaultDocOut>(API_VAULT_BY_PATH_ENDPOINT, merged);
}
